From a1c0440e21104e4dc60b564a1a5f05accfecdbb3 Mon Sep 17 00:00:00 2001
From: Christian Hoene <christian.hoene@symonics.com>
Date: Sat, 24 Jan 2026 13:37:13 +0100
Subject: [PATCH] optimized code of btree loop

---
 src/hdf/btree.c | 96 +++++++++++++++++++++++++++++++++++--------------
 1 file changed, 69 insertions(+), 27 deletions(-)

diff --git b/src/hdf/btree.c a/src/hdf/btree.c
index 674017a..fb37dbc 100644
--- b/src/hdf/btree.c
+++ a/src/hdf/btree.c
@@ -212,8 +212,8 @@ void btreeFree(struct BTREE *btree) { free(btree->records); }
 
 int treeRead(struct READER *reader, struct DATAOBJECT *data) {
 
-  int d, p, b = 0, c[HDF_MAX_DIMENSIONALITY];
-  int i, j, err, olen, elements, size, e;
+  int i, j, err, olen, elements, size, x, y, z, w, b, e, dy, dz, sx, sy, sz, dw,
+      sw, k;
   char *input, *output;
 
   uint8_t node_type, node_level;
@@ -256,6 +256,13 @@ int treeRead(struct READER *reader, struct DATAOBJECT *data) {
   elements = 1;
   for (j = 0; j < data->ds.dimensionality; j++)
     elements *= data->datalayout_chunk[j];
+  dy = data->datalayout_chunk[1];
+  dz = data->datalayout_chunk[2];
+  dw = data->datalayout_chunk[3];
+  sx = data->ds.dimension_size[0];
+  sy = data->ds.dimension_size[1];
+  sz = data->ds.dimension_size[2];
+  sw = data->ds.dimension_size[3];
   size = data->datalayout_chunk[data->ds.dimensionality];
 
   mylog("elements %d size %d\n", elements, size);
@@ -317,42 +324,77 @@ int treeRead(struct READER *reader, struct DATAOBJECT *data) {
         return MYSOFA_INVALID_FORMAT; // LCOV_EXCL_LINE
       }
 
-      for (d = 0; d < data->ds.dimensionality; d++) {
-        c[d] = 0;
-      }
+      switch (data->ds.dimensionality) {
+      case 1:
+        for (i = 0; i < olen; i++) {
+          b = i / elements;
+          x = i % elements + start[0];
+          if (x < sx) {
 
-      for (i = 0; i < olen; i++) {
+            j = x * size + b;
+            if (j >= 0 && j < data->data_len) {
+              ((char *)data->data)[j] = output[i];
+            }
+          }
+        }
+        break;
 
-        // calculate and check destination pointer
-        p = 0;
-        for (d = 0; d < data->ds.dimensionality; d++) {
-          if (c[d] + start[d] >= data->ds.dimension_size[d]) {
-            p = -1;
-            break;
+      case 2:
+        for (i = 0; i < olen; i++) {
+          b = i / elements;
+          k = i % elements;
+          y = k % dy + start[1];
+          k = k / dy;
+          x = k + start[0];
+          if (y < sy && x < sx) {
+            j = ((x * sy + y) * size) + b;
+            if (j >= 0 && j < data->data_len) {
+              ((char *)data->data)[j] = output[i];
+            }
           }
-          p = p * data->ds.dimension_size[d] + c[d] + start[d];
         }
+        break;
 
-        // copy data if within bounds
-        if (p >= 0) {
-          p = p * size + b;
-          if (p < data->data_len) {
-            ((char *)data->data)[p] = output[i];
+      case 3:
+        for (i = 0; i < olen; i++) {
+          b = i / elements;
+          k = i % elements;
+          z = k % dz + start[2];
+          k = k / dz;
+          y = k % dy + start[1];
+          k = k / dy;
+          x = k + start[0];
+          if (z < sz && y < sy && x < sx) {
+            j = ((x * sy + y) * sz + z) * size + b;
+            if (j >= 0 && j < data->data_len) {
+              ((char *)data->data)[j] = output[i];
+            }
           }
         }
+        break;
 
-        // increase source coordinates
-        for (d = data->ds.dimensionality - 1; d >= 0; d--) {
-          c[d]++;
-          if (c[d] < data->datalayout_chunk[d]) {
-            break;
-          } else {
-            c[d] = 0;
-            if (d == 0) { // at last, increase bytes
-              b++;
+      case 4:
+        for (i = 0; i < olen; i++) {
+          b = i / elements;
+          k = i % elements;
+          w = k % dw + start[3];
+          k = k / dw;
+          z = k % dz + start[2];
+          k = k / dz;
+          y = k % dy + start[1];
+          k = k / dy;
+          x = k + start[0];
+          if (z < sz && y < sy && x < sx && w < sw) {
+            j = (((x * sy + y) * sz + z) * sw + w) * size + b;
+            if (j >= 0 && j < data->data_len) {
+              ((char *)data->data)[j] = output[i];
             }
           }
         }
+        break;
+      default:
+        mylog("invalid dim\n");       // LCOV_EXCL_LINE
+        return MYSOFA_INTERNAL_ERROR; // LCOV_EXCL_LINE
       }
 
       if (mysofa_seek(reader, store, SEEK_SET) < 0) {
-- 
2.53.0

